qadence_protocols/measurements/protocols.py

```diff
@@ -1,115 +1,69 @@
-
-        shadow_size = self.options.get("shadow_size", None)
-        if shadow_size is None:
-            raise KeyError("Robust Shadow protocol requires an option 'shadow_size' of type 'int'.")
-        shadow_groups = self.options.get("shadow_groups", None)
-        if shadow_groups is None:
-            raise KeyError("Shadow protocol requires an option 'shadow_groups' of type 'int'.")
-
-        calibration = self.options.get("calibration", None)
+class Measurements(Protocol):
+    """Define a measurement protocol using a `MeasurementManager`.
 
-        return {
-            "shadow_size": shadow_size,
-            "shadow_groups": shadow_groups,
-            "calibration": calibration,
-        }
+    Possible options are available via the `MeasurementProtocols` type.
+    The `MeasurementManager` instance has a verify_options
 
-
-@dataclass
-class Measurements(Protocol):
-    TOMOGRAPHY = "tomography"
-    SHADOW = "shadow"
-    ROBUST_SHADOW = "robust_shadow"
+    Attributes:
+        protocol (str): Protocol name.
+        options (dict): Options to run protocol.
+    """
 
     def __init__(self, protocol: str, options: dict = dict()) -> None:
-        verified_options = MeasurementOptions(protocol, options).verify_options()
+        try:
+            module = importlib.import_module(PROTOCOL_TO_MODULE[protocol][0])
+            proto_class = getattr(module, PROTOCOL_TO_MODULE[protocol][1])
+        except (KeyError, ModuleNotFoundError, ImportError) as e:
+            raise type(e)(f"Failed to import Mitigations due to {e}.")
+        self.manager: MeasurementManager = proto_class(data=None, options=options)
+        verified_options = self.manager.validate_options()
         super().__init__(protocol, verified_options)
 
     def __call__(
         self,
         model: QuantumModel,
         param_values: dict[str, Tensor] = dict(),
+        state: Tensor | None = None,
         return_expectations: bool = True,
     ) -> Tensor:
-        """Compute expectation values via measurements.
+        """Compute measurements or expectation values via measurements.
 
         Args:
             model (QuantumModel): Model to evaluate.
             param_values (dict[str, Tensor], optional): Parameter values. Defaults to dict().
+            state (Tensor | None, optional): Input state. Defaults to None.
 
         Returns:
             Tensor: Expectation values.
         """
-        try:
-            module = importlib.import_module(PROTOCOL_TO_MODULE[self.protocol])
-        except (KeyError, ModuleNotFoundError, ImportError) as e:
-            raise type(e)(f"Failed to import Mitigations due to {e}.")
 
         conv_observables = model._observable
         observables = [obs.abstract for obs in conv_observables]
 
         # Partially pass the options and observable.
-        compute_fn = "compute_expectation" if return_expectations else "compute_measurements"
+        compute_fn = "expectation" if return_expectations else "measure"
```

Comment:
hm, it feels like there are two distinct return types here. For the sake of simplicity, I would choose one or the other but not let a method potentially return either as it makes its use quite confusing. That could be another PR.

Reply:
Removed